course-mw/enrolment-actor/src/main/scala/org/sunbird/enrolments/CourseEnrolmentActor.scala

```diff
@@ -590,7 +589,27 @@ class CourseEnrolmentActor @Inject()(@Named("course-batch-notification-actor") c
 
     def getCoursesForProgramAndEnrol(request: Request, programId: String, userId: String, batchId: String) = {
         val redisKey = s"$programId:$programId:childrenCourses"
-        val childrenNodes: List[String] = cacheUtil.getList(redisKey, redisCollectionIndex)
+        var childrenNodes: List[String] = List.empty[String] //cacheUtil.getList(redisKey, redisCollectionIndex)
+        var contentDataForProgram: java.util.List[java.util.Map[String, AnyRef]] = null
+        var isFoundInCache: Boolean = true
+        if (childrenNodes.isEmpty) {
+            isFoundInCache = false
+            contentDataForProgram = contentHierarchyDao.getContentChildren(request.getRequestContext, programId)
+            if (CollectionUtils.isNotEmpty(contentDataForProgram)) {
+                childrenNodes = contentDataForProgram.asScala.toStream
+                  .flatMap { childNode =>
+                      Option(childNode.get(JsonKey.IDENTIFIER)).map(_.asInstanceOf[String])
+                  }.toList
+            } else {
+                ProjectCommonException.throwClientErrorException(ResponseCode.missingHierarchyContentDataForProgram);
+            }
+            if (!checkCourseBatchInfo(childrenNodes, request.getRequestContext)) {
+                ProjectCommonException.throwClientErrorException(ResponseCode.courseDoesNotHaveBatch);
+            }
+        }
+        if (isFoundInCache && !checkCourseBatchInfo(childrenNodes, request.getRequestContext)) {
+            ProjectCommonException.throwClientErrorException(ResponseCode.courseDoesNotHaveBatch);
+        }
```

Comment:
Common mistakes made - 
checkCourseBatchInfo -- this method called multiple times unnecessarily 
The call made to get the courseBatchDao.readFirstAvailableBatch -- is not store if the value exist -- which is the case of 99.9% -- hence another call is made from inside method nrollProgramCourses 
cacheUtil is been removed -- hence one more additional cassandra call made everytime 

the below is the pesudo code for this implementation... 

childrenNodes -- get from cache.
//Create new Map named courseBatchMap (key is courseId, value is batchId) -- to store childCoureseId and childCourseBatchId

if (!childrenNodes.isEmpty) {
	//read content data. Get primaryCategory
	//Check primaryCagtegory  is allowed - if not throw error.
	//Read course batch -- if not available throw error.
	//No errors at this point.. save the courseId and batchId into courseBatchMap
} else {
	//Get hierarchy and get child nodes
	//Loop throw child nodes.
	   //Get content data. Get primaryCategory
	   //Check primaryCategory is allowed - if not throw error.
	   //Read course batch -- if not avaliable throw error.
	   //No errors at this point.. save the courseId and batchId into courseBatchMap
}

//Iterate throw courseBatchMap
for(courseBatchObj : courseBatchMap) {
   //Set the courseId of the current iteration into request obj.
   //call enrollProgramCourse(request, batchId)
}

enrollProgramCourses -- update method signature to use batchId -- use the batchId and remove the code which is getting batchData

Reply:
@shankaragoudab  Please use courseBatch Object for passing into the enrollPorgram as we are using in validateEnrollement and notifyUser